server/core/include/irods_logger.tpp

```diff
@@ -0,0 +1,217 @@
+template <log::level Level>
+class log::logger<Category>::impl
+{
+public:
+    template <typename T>
+    using is_iterable = decltype(std::begin(std::declval<std::decay_t<T>>()));
+
+    void operator()(const std::string& _msg) const
+    {
+        (*this)({{"msg", _msg}});
+    }
+
+    void operator()(std::initializer_list<log::key_value> _list) const noexcept
+    {
+        (*this)(std::begin(_list), std::end(_list));
+    }
+
+    template <typename Container,
+              typename = is_iterable<Container>>
+    void operator()(const Container& _container) const noexcept
+    {
+        (*this)(std::begin(_container), std::end(_container));
+    }
+
+    template <typename ForwardIt>
+    void operator()(ForwardIt _first, ForwardIt _last) const noexcept
+    {
+        if (!should_log())
+        {
+            return;
+        }
+
+        const auto msg = to_json_string(logger_config<Category>::name, _first, _last);
+
+        if constexpr (Level == level::trace)
+        {
+            log_->trace(msg);
+        }
+        else if constexpr (Level == level::debug)
+        {
+            log_->debug(msg);
+        }
+        else if constexpr (Level == level::info)
+        {
+            log_->info(msg);
+        }
+        else if constexpr (Level == level::warn)
+        {
+            log_->warn(msg);
+        }
+        else if constexpr (Level == level::error)
+        {
+            log_->error(msg);
+        }
+        else if constexpr (Level == level::critical)
+        {
+            log_->critical(msg);
+        }
+
+        append_to_r_error_stack(_first, _last);
+    }
+
+private:
+    impl() = default;
+
+    impl(const impl&) = delete;
+    impl& operator=(const impl&) = delete;
+
+    bool should_log() const noexcept
+    {
+        return Level >= logger_config<Category>::level;
+    }
+
+    template <typename ForwardIt>
+    std::string to_json_string(const std::string& _src, ForwardIt _first, ForwardIt _last) const
+    {
+        using json = nlohmann::json;
+        using container = std::unordered_map<std::string, std::string>;
+
+        container data(_first, _last);
+        data["src"] = _src;
+
+        return json(data).dump();
+    }
+
+    template <typename ForwardIt>
+    void append_to_r_error_stack(ForwardIt _first, ForwardIt _last) const
+    {
+        if (error_ && write_to_error_object_)
+        {
+            std::for_each(_first, _last, [](const auto& _key_value) {
+                std::string msg = _key_value.first;
+                msg += ": ";
+                msg += _key_value.second;
+
+                // Force addRErrorMsg() to skip prefixing the message with 'Level N:'.
+                // When the output is printed on the client-side, it will very close to
+                // the log file.
+                constexpr int disable_level_prefix = STDOUT_STATUS;
+
+                if (const auto ec = addRErrorMsg(error_, disable_level_prefix, msg.c_str()); ec)
```

Comment:
I like this style, we should make sure to do it. Even when we are evaluating the we initialize, explicitly evaluating it afterwards makes it more consistent and clearer to read.

Reply:
I totally agree.